File: scripts/ui.py
#!/usr/bin/env python3

import rospy
from turtlesim.srv import Spawn
from geometry_msgs.msg import Twist
from assignment1_rt.msg import Vel

pub1 = rospy.Publisher('/turtle1/cmd_vel', Twist, queue_size=10)
pub2 = rospy.Publisher('/turtle2/cmd_vel', Twist, queue_size=10)
pub3 = rospy.Publisher('/turtle1/vel', Vel, queue_size=10)
pub4 = rospy.Publisher('/turtle2/vel', Vel, queue_size=10)

def spawn_turtle():
    x = rospy.get_param("/turtle2_pose/x")
    y = rospy.get_param("/turtle2_pose/y")
    theta = rospy.get_param("/turtle2_pose/theta")

    rospy.wait_for_service('/spawn')
    try:
        spawn_turtle = rospy.ServiceProxy('/spawn', Spawn)
        request = spawn_turtle(x, y, theta,"turtle2")
        rospy.loginfo("Successfully spawned turtle2")
    except rospy.ServiceException as e:
        rospy.logerr(f"Service call failed: {e}")

def get_user_input():
    # Get user input for turtle selection and velocity 
    turtle = input("Enter turtle to control:\n"
                   "- turtle1\n"
                   "- turtle2\n")
    linear_vel = float(input("Enter linear velocity: "))
    angular_vel = float(input("Enter angular velocity: "))
    return turtle, linear_vel, angular_vel

def control_turtle(turtle, linear_vel, angular_vel):
    #pub = rospy.Publisher(f'/{turtle}/cmd_vel', Twist, queue_size=10)
    #while pub.get_num_connections() == 0:
    #    rospy.sleep(0.1)  # Wait for connection
    
    pub = pub1 if turtle == "turtle1" else pub2
    pub_vel = pub3 if turtle == "turtle1" else pub4

    twist = Twist()
    vel = Vel()
    twist.linear.x = linear_vel
    vel.linear_velocity = linear_vel
    twist.linear.y = 0
    twist.linear.z = 0
    twist.angular.x = 0
    twist.angular.y = 0
    twist.angular.z = angular_vel
    vel.angular_velocity = angular_vel
    pub.publish(twist)
    pub_vel.publish(vel)

    # Sleep for 1 second
    rospy.sleep(rospy.Duration(1)) 
    
    # Stop after duration
    twist.linear.x = 0
    twist.angular.z = 0
    vel.linear_velocity = 0
    vel.angular_velocity = 0
    pub.publish(twist)
    pub_vel.publish(vel)

    # #rate = rospy.Rate(1)
    # elapsed_time = 0
    # start_time = rospy.get_rostime()
    # while elapsed_time < 1.0:
    #     # Calculate elapsed time in seconds
    #     pub.publish(twist)
    #     elapsed_time = (rospy.get_rostime() - start_time).to_sec()
    #     #rate.sleep()

def main():
    rospy.init_node('ui_node',anonymous=True)
    #rate = rospy.Rate(1)

    spawn_turtle()  # Spawn turtle2 at startup
    while not rospy.is_shutdown():
        turtle, linear_vel, angular_vel = get_user_input()
        if turtle in ["turtle1", "turtle2"]:
            control_turtle(turtle, linear_vel, angular_vel)
        else:
            rospy.logwarn("Invalid turtle name. Choose either 'turtle1' or 'turtle2'.")
        #rate.sleep()

if __name__ == "__main__":
    main()

File: src/ui.cpp
#include "ros/ros.h"
#include "turtlesim/Spawn.h"
#include "geometry_msgs/Twist.h"
#include "assignment1_rt/Vel.h"
#include <iostream>
#include <string>

ros::Publisher pub1;
ros::Publisher pub2;
ros::Publisher pub3;
ros::Publisher pub4;

struct TurtleInput {
    std::string turtle;
    float linear_vel;
    float angular_vel;
};

TurtleInput getUserInput() {
    // Get user input for turtle selection and velocity
	TurtleInput input;

	std::cout << "Enter turtle to control:\n"
              << "- turtle1\n"
              << "- turtle2\n";
    std::cin >> input.turtle;

    std::cout << "Enter linear velocity: ";
    std::cin >> input.linear_vel;

    std::cout << "Enter angular velocity: ";
    std::cin >> input.angular_vel;
    
    return input;
}

void control_turtle(const std::string turtle, float linear_vel,float angular_vel){
    // Choose the correct publisher based on the turtle
	ros::Publisher pub = (turtle == "turtle1") ? pub1 : pub2;
	ros::Publisher pub_vel = (turtle == "turtle1") ? pub3 : pub4;

	// Create a Twist message and set linear and angular velocities
	geometry_msgs::Twist twist;
	assignment1_rt::Vel vel;

	twist.linear.x = linear_vel;
	vel.linear_velocity = linear_vel;
	twist.linear.y = 0;
	twist.linear.z = 0;
	twist.angular.x = 0;
	twist.angular.y = 0;
	twist.angular.z = angular_vel;
	vel.angular_velocity = angular_vel;


	// Publish the Twist message
	pub.publish(twist);
	pub_vel.publish(vel);
	
	ros::Duration(1.0).sleep(); // sleep for 1 second
	
	// Stop after duration
	twist.linear.x = 0;
    twist.angular.z = 0;
	vel.linear_velocity = 0;
	vel.angular_velocity= 0;
    pub.publish(twist);
	pub_vel.publish(vel);

}

int main(int argc, char **argv){
    ros::init(argc, argv, "ui");
    ros::NodeHandle nh;
	// ros::Rate loop_rate(1); 

	pub1 = nh.advertise<geometry_msgs::Twist>("/turtle1/cmd_vel", 10);
    pub2 = nh.advertise<geometry_msgs::Twist>("/turtle2/cmd_vel", 10);

	pub3 = nh.advertise<assignment1_rt::Vel>("/turtle1/vel", 10);
    pub4 = nh.advertise<assignment1_rt::Vel>("/turtle2/vel", 10);

    ros::ServiceClient spawn_turtle = nh.serviceClient<turtlesim::Spawn>("/spawn");
	turtlesim::Spawn srv1;
	srv1.request.x = 2.0;
	srv1.request.y = 3.0;
 	srv1.request.theta = 0.0;
 	srv1.request.name = "turtle2";
 	spawn_turtle.waitForExistence();
	if (spawn_turtle.call(srv1))
		ROS_INFO("Successfully spawned turtle2"); 
	else
	{
		ROS_ERROR("Failed to call service"); 
		return 1;
	}	 

	while (ros::ok())
	{
		TurtleInput input;
		input = getUserInput();
		if (input.turtle == "turtle1" || input.turtle == "turtle2") {
            control_turtle(input.turtle, input.linear_vel, input.angular_vel);
		}
        else{
			ROS_WARN("Invalid turtle name. Choose either 'turtle1' or 'turtle2'.");
		}
		ros::spinOnce();
        // loop_rate.sleep();
	}
	return 0;
}